### PR TITLE
Add all standard TXO types to bitcoin-tx

### PR DESCRIPTION
**Pull request by jnewbery**
bitcoin-tx can currently create transactions with the following TXO types:
- Pay to Pub Key Hash
- null data (OP_RETURN)
- Pay to Script Hash

This PR enhances bitcoin-tx so all remaining standard TXO types can be created:
- Pay to Pub Key
- Multi-sig
  - bare multi-sig
  - multi-sig in Pay To Script Hash
  - multi-sig in Pay to Witness Script Hash
  - multi-sig in Pay to Witness Script Hash, wrapped in P2SH
- Pay to Witness Pub Key Hash
  - Pay to Witness Pub Key Hash, wrapped in P2SH
- Pay to Witness Script Hash
  - Pay to Witness Script Hash, wrapped in P2SH

The PR also includes new test cases for all tx types.

### CODE REVIEW

**Comment by fanquake** [comment]
Windows builds are failing:
```
make[2]: Entering directory `/home/travis/build/bitcoin/bitcoin/build/src'
  CXX      bitcoin_tx-bitcoin-tx.o
../../src/bitcoin-tx.cpp: In function ‘void MutateTxAddOutMultiSig(CMutableTransaction&, const string&)’:
../../src/bitcoin-tx.cpp:309:5: error: ‘uint’ was not declared in this scope
     uint required = std::stoul(vStrInputParts[1]);
     ^
../../src/bitcoin-tx.cpp:309:10: error: expected ‘;’ before ‘required’
     uint required = std::stoul(vStrInputParts[1]);
          ^
../../src/bitcoin-tx.cpp:312:10: error: expected ‘;’ before ‘numkeys’
     uint numkeys = std::stoul(vStrInputParts[2]);
          ^
../../src/bitcoin-tx.cpp:315:33: error: ‘numkeys’ was not declared in this scope
     if (vStrInputParts.size() < numkeys + 3)
                                 ^
../../src/bitcoin-tx.cpp:318:9: error: ‘required’ was not declared in this scope
     if (required < 1 || required > 20 || numkeys < 1 || numkeys > 20 || numkeys < required)
         ^
../../src/bitcoin-tx.cpp:318:42: error: ‘numkeys’ was not declared in this scope
     if (required < 1 || required > 20 || numkeys < 1 || numkeys > 20 || numkeys < required)
                                          ^
../../src/bitcoin-tx.cpp:324:14: error: expected ‘;’ before ‘pos’
     for(uint pos = 1; pos <= numkeys; pos++) {
              ^
../../src/bitcoin-tx.cpp:324:23: error: ‘pos’ was not declared in this scope
     for(uint pos = 1; pos <= numkeys; pos++) {
                       ^
../../src/bitcoin-tx.cpp:324:30: error: ‘numkeys’ was not declared in this scope
     for(uint pos = 1; pos <= numkeys; pos++) {
                              ^
../../src/bitcoin-tx.cpp:334:34: error: ‘numkeys’ was not declared in this scope
     if (vStrInputParts.size() == numkeys + 4) {
                                  ^
../../src/bitcoin-tx.cpp:344:49: error: ‘required’ was not declared in this scope
     CScript scriptPubKey = GetScriptForMultisig(required, pubkeys);
                                                 ^
make[2]: *** [bitcoin_tx-bitcoin-tx.o] Error 1
make[2]: Leaving directory `/home/travis/build/bitcoin/bitcoin/build/src'
make[1]: *** [check-recursive] Error 1
make[1]: Leaving directory `/home/travis/build/bitcoin/bitcoin/build/src'
make: *** [check-recursive] Error 1
```

**Comment by paveljanik** [comment]
Concept ACK

New warning:
```
bitcoin-tx.cpp:283:25: warning: declaration shadows a local variable [-Wshadow]
        CBitcoinAddress addr(scriptPubKey);
                        ^
bitcoin-tx.cpp:265:21: note: previous declaration is here
    CBitcoinAddress addr(scriptPubKey);
                    ^
1 warning generated.
```

**Comment by jnewbery** [comment]
Thanks @paveljanik . Fixed in squashed commit.

**Comment by paveljanik** [comment]
`-Wshadow` clean now, thanks!

I'll read the changes later tonight.

File: src/bitcoin-tx.cpp [hunk 1]
```diff
@@ -75,8 +75,16 @@ static bool AppInitRawTx(int argc, char* argv[])
...
+        strUsage += HelpMessageOpt("outpubkey=VALUE:PUBKEY[:FLAGS]", _("Add pay-to-pubkey output to TX") + ". " +
+            _("Optionally add the \"W\" flag to produce a pay-to-witness-pubkey-hash output") + ". " +
+            _("Optionally add the \"S\" flag to wrap the output in a pay-to-script-hash."));
         strUsage += HelpMessageOpt("outdata=[VALUE:]DATA", _("Add data-based output to TX"));
-        strUsage += HelpMessageOpt("outscript=VALUE:SCRIPT", _("Add raw script output to TX"));
+        strUsage += HelpMessageOpt("outscript=VALUE:SCRIPT[:FLAGS]", _("Add raw script output to TX") + ". " +
+            _("Optionally add the \"W\" flag to produce a pay-to-witness-script-hash output") + ". " +
+            _("Optionally add the \"S\" flag to wrap the output in a pay-to-script-hash."));
+        strUsage += HelpMessageOpt("outmultisig=VALUE:REQUIRED:PUBKEYS:PUBKEY1:PUBKEY2:....[:FLAGS]", _("Add Pay To n-of-m Multi-sig output to TX. n = REQUIRED, m = PUBKEYS") + ". " +
+            _("Optionally add the \"W\" flag to produce a pay-to-witness-script-hash output") + ". " +
```
Comment (paveljanik) [inline src/bitcoin-tx.cpp]: Why do you sometimes translate dot at the end of sentence and sometimes not?

File: src/bitcoin-tx.cpp [hunk 2]
```diff
@@ -215,33 +223,142 @@ static void MutateTxAddInput(CMutableTransaction& tx, const string& strInput)
 
 static void MutateTxAddOutAddr(CMutableTransaction& tx, const string& strInput)
 {
-    // separate VALUE:ADDRESS in string
-    size_t pos = strInput.find(':');
-    if ((pos == string::npos) ||
-        (pos == 0) ||
-        (pos == (strInput.size() - 1)))
-        throw runtime_error("TX output missing separator");
+    // Seperate into VALUE:ADDRESS
```
Comment (paveljanik) [inline src/bitcoin-tx.cpp]: Separate (e -> a)?

File: src/bitcoin-tx.cpp [hunk 3]
```diff
@@ -215,33 +223,142 @@ static void MutateTxAddInput(CMutableTransaction& tx, const string& strInput)
...
+    vector<string> vStrInputParts;
+    boost::split(vStrInputParts, strInput, boost::is_any_of(":"));
```
Comment (paveljanik) [inline src/bitcoin-tx.cpp]: This adds another dependency on boost. But as we already use it in this file, it is OK.

File: src/bitcoin-tx.cpp (hunk 2 above)
Comment (czzarr) [inline src/bitcoin-tx.cpp]: fixed

File: src/bitcoin-tx.cpp (hunk 1 above)
Comment (czzarr) [inline src/bitcoin-tx.cpp]: for some reason I get a compile error when I put the dot in the sentence:
```
bitcoin-tx.cpp:79:13: error: expected ')'
            _("Optionally add the \"W\" flag to produce a pay-to-witness-pubkey-hash output") + ". " +
            ^
bitcoin-tx.cpp:78:35: note: to match this '('
        strUsage += HelpMessageOpt("outpubkey=VALUE:PUBKEY[:FLAGS]", _("Add pay-to-pubkey output to TX. ")
                                  ^
1 error generated.
```

File: src/bitcoin-tx.cpp [hunk 4]
```diff
@@ -215,33 +223,142 @@ static void MutateTxAddInput(CMutableTransaction& tx, const string& strInput)
...
     tx.vout.push_back(txout);
 }
 
+static void MutateTxAddOutPubKey(CMutableTransaction& tx, const string& strInput)
+{
+    // Seperate into VALUE:PUBKEY[:FLAGS]
+    vector<string> vStrInputParts;
+    boost::split(vStrInputParts, strInput, boost::is_any_of(":"));
+
+    // Extract and validate VALUE
```
Comment (czzarr) [inline src/bitcoin-tx.cpp]: done in https://github.com/bitcoin/bitcoin/pull/8883/commits/79cd9275db714603f4a3c55360b3d52a945d95da

File: src/bitcoin-tx.cpp [hunk 5]
```diff
@@ -215,33 +231,133 @@ static void MutateTxAddInput(CMutableTransaction& tx, const string& strInput)
...
+    // Extract and validate FLAGS
+    bool bSegWit = false;
+    bool bScriptHash = false;
+    if (vStrInputParts.size() == 3) {
+        string flags = vStrInputParts[2];
+        bSegWit = (flags.find("W") != string::npos);
+        bScriptHash = (flags.find("S") != string::npos);
+    }
+
+    if (bSegWit) {
```
Comment (jonasschnelli) [inline src/bitcoin-tx.cpp]: Is there a reason for `bSegWit` and not directly use `(flags.find("W") != string::pos)`?

File: src/bitcoin-tx.cpp [hunk 6]
```diff
@@ -215,33 +231,133 @@ static void MutateTxAddInput(CMutableTransaction& tx, const string& strInput)
...
+        // Call GetScriptForWitness() to build a P2WSH scriptPubKey
+        scriptPubKey = GetScriptForWitness(scriptPubKey);
+    }
+    if (bScriptHash) {
```
Comment (jonasschnelli) [inline src/bitcoin-tx.cpp]: maybe a `else if`?

File: src/bitcoin-tx.cpp [hunk 7]
```diff
@@ -215,33 +231,133 @@ static void MutateTxAddInput(CMutableTransaction& tx, const string& strInput)
...
+    boost::split(vStrInputParts, strInput, boost::is_any_of(":"));
+
+    // Extract and validate VALUE
+    CAmount value = ExtractAndValidateValue(vStrInputParts[0]);
+
+    // Extract and validate PUBKEY
+    CPubKey pubkey(ParseHex(vStrInputParts[1]));
+    if (!pubkey.IsFullyValid())
+        throw runtime_error("invalid TX output pubkey");
+    CScript scriptPubKey = GetScriptForRawPubKey(pubkey);
```
Comment (jonasschnelli) [inline src/bitcoin-tx.cpp]: Does this result in supporting pure `P2PK` (not `P2PKH`)?

**Comment by jonasschnelli** [comment]
Concept ACK.

**Review by jonasschnelli — CHANGES_REQUESTED** [review]

File: src/bitcoin-tx.cpp (hunk 7 above)
Comment (jnewbery) [inline src/bitcoin-tx.cpp]: Yes - without either flag, the TXO will be a P2PK.

- With just the `H` flag, the TXO is a P2SH
- With just the `W` flag, the TXO is a P2WPK
- With both the `H` and `W` flags, the TXO is a P2WPK wrapped in a P2SH

File: src/bitcoin-tx.cpp (hunk 5 above)
Comment (jnewbery) [inline src/bitcoin-tx.cpp]: See below. The flags are not exclusive. I need to put the `flags.find()` test inside the  `(vStrInputParts.size() == 3)` so I may as well parse for both flags at the same point and store the result in variables.

File: src/bitcoin-tx.cpp (hunk 6 above)
Comment (jnewbery) [inline src/bitcoin-tx.cpp]: They're not exclusive. If both flags are used, we output a P2WPKH wrapped in P2SH.

**Comment by jnewbery** [comment]
Rebased. @jonasschnelli - are you happy with my responses above?

**Comment by jonasschnelli** [comment]
@jnewbery: Yes.
utACK, will test now...

**Comment by jonasschnelli** [comment]
Travis is failing on the new tests:
```
Running test/bitcoin-util-test.py...
make[4]: Entering directory `/home/travis/build/bitcoin/bitcoin/build/src'
Output data mismatch for txcreatescript1.json
make[3]: *** [check-local] Error 1
make[3]: *** Waiting for unfinished jobs....
```

**Comment by jnewbery** [comment]
I've fixed the failing testcases and squashed the commits. @jonasschnelli - please let me know if you have any further feedback for this PR.

**Review by jonasschnelli — APPROVED** [review]

**Comment by jonasschnelli** [comment]
slightly tested ACK e5fecbf64d1dad360490294d1aad79497dcb1de7.
Someone should verify the unit tests fixtures.

**Comment by laanwj** [comment]
utACK.
Needs rebase though.

**Comment by jnewbery** [comment]
rebased

**Comment by jnewbery** [comment]
Rebased. I had to do a couple of large merges (removing std namespace and updating default tx version). @laanwj - let me know if there's anything I can do to help get this merged into master.

**Pull request merged**